### ISSUE TITLE
Pulumi creates empty zip

### ISSUE THREAD

**Issue by duongdominhchau** [pulumi/pulumi-aws#3530]
### What happened?

Non-empty zip is being uploaded as empty zip on S3. This zip file is created using Python `zipfile`

```py
with ZipFile("a.zip", "w", compression=ZIP_DEFLATED) as zf:
    for p in ...:
        zf.writestr(ZipInfo(path), p.read_bytes())
```

### Example

1. Clone https://github.com/duongdominhchau/empty-zip-repro
2. Change the bucket name here and run `pulumi up`, an empty zip (22 bytes) is created on S3, but the source file is not empty.

### Output of `pulumi about`

$ pulumi about                                                                                                                                                                                                                       main -> origin/main
CLI
Version      3.106.0
Go Version   go1.22.0
Go Compiler  gc

Plugins
NAME    VERSION
aws     6.24.0
python  unknown

Host
OS       arch
Version
Arch     x86_64

This project is written in python: executable='/home/chau/test/.direnv/python-3.12/bin/python3' version='3.12.2'

Current Stack: organization/test/dev

TYPE                                  URN
pulumi:pulumi:Stack                   urn:pulumi:dev::test::pulumi:pulumi:Stack::test-dev
pulumi:providers:aws                  urn:pulumi:dev::test::pulumi:providers:aws::default_6_24_0
aws:s3/bucketObjectv2:BucketObjectv2  urn:pulumi:dev::test::aws:s3/bucketObjectv2:BucketObjectv2::a


Found no pending operations associated with dev

Backend
Name           archlinux
URL            file://~
User           chau
Organizations
Token type     personal

Dependencies:
NAME        VERSION
pip         24.0
pulumi_aws  6.24.0
pyright     1.1.352

Pulumi locates its logs in /tmp by default

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by duongdominhchau** [comment]
Sorry for not being clear enough about this. I tried uploading the zip file in the repo above to 2 different buckets using Pulumi, each bucket is in a different AWS account and both have this issue. 

The code used for creating the zip file has some other filtering logic to determine what files in the directory is included in the zip, but I think that part is irrelevant so I reduced that piece of code to the core part only.

I also tried extracting it and use `7z` to create another zip file from the extracted content and it works. I tried removing a file from that and then use the same piece of code to create the zip and it works too.

I'm currently busy with some stuff, I'll get back to this later to provide more detailed information. For now, please try changing the bucket name in my repo to your bucket and run `pulumi up`, I wrote a `__main__.py` to upload the problematic `a.zip` to S3. Thanks.

**Comment by duongdominhchau** [comment]
Today I finally have time to have another look at this, but then the problem disappeared. When I created this issue, I can reproduce the problem with different buckets, so I thought this specific zip file is the problem, but with what happened today it turned out I was wrong.

In the past, I experienced this issue a few times. They all happened when I create a zip for Lambda deployment. AWS API returns an error to Pulumi saying the zip is empty but checking the file locally it is clearly non-empty. Deleting the file manually then `pulumi up --refresh --yes` doesn't help, so I moved that zip file to a separate repo and try deploying it to another bucket, still get the error. That's what makes me thought the zip file is the problem. Today I updated the Pulumi script to deploy into a bucket created by Pulumi, so I go ahead and delete the old bucket, then the problem disappears.

Given that it is not reproducible anymore, I'm closing this ticket now, apologize for the long wait. Hopefully the next time it happens I can collect more information to write a better reproducible example.
